automatic-cluster-upgrades/README.md

```diff
@@ -72,25 +74,29 @@ Example resource:
 apiVersion: 
 kind: ClusterUpgrade
 metadata:
-  name: upgrade-to-release-v20
+  name: upgrade-to-cluster-openstack-v1.20
   namespace: default
 spec:
-  clusterRef:
-    apiVersion: cluster.x-k8s.io/v1alpha3
-    kind: Cluster
-    name: foo01
+  matchers:
+  - name: cluster_id
+    value: mgab6
+    isRegex: false
+  - name: dev_clusters
+    key: stage
+    value: dev
+    isRegex: true
```

Comment:
@piontec moving the discussion here. This is a proposal but I got your point of making this object tracked by Git so the GitOps workflow is not broken. 

Two things:

1) We could change `ClusterUpgrade` object to be tied to a single cluster (1:1 relation). That way the `status` would reflect the result of the upgrade and can help us to track upgrades, and at same time can be added to Git repo as part of GitOps workflow

2) The Cluster Upgrade controller instead of creating `ClusterUpgrade` resource in MCs can do PRs to your Git repo in case of GitOps enabled. Otherwise can default to MC API

Reply:
I like the simplicity of a 1:1 relationship between `ClusterUpgrade` object and `Cluster` however I also feel there is scope for requirements gathering on this point - how do customers *want* to implement upgrades? Is a 1:1 mapping desired or would they want to upgrade by `Organization` - I think there needs to be a balance here between simplicity and desirability here.

Reply:
Irrespective of `GitOps` process, I also think that we need to track the current deployed version for an individual cluster, especially so in larger customers where many versions might be in play at any one time. Tracking of deployed versions would be important in the event of DR or cluster rebuilds where deploying an upgraded version might have detrimental impact on customers ability to operate their workloads.

Reply:
> I like the simplicity of a 1:1 relationship between ClusterUpgrade object and Cluster however I also feel there is scope for requirements gathering on this point - how do customers want to implement upgrades? Is a 1:1 mapping desired or would they want to upgrade by Organization - I think there needs to be a balance here between simplicity and desirability.

I do think it can be tackled with higher abstraction. For example, in GitOps we will charge a base and it will run all attached clusters, and it is defined by the customer 💪 . In other scenarios, we can rely on a higher entity but having the `ClusterUpgrade` entity tied to a single cluster will help in most cases IMO

> Irrespective of GitOps process, I also think that we need to track the current deployed version for an individual cluster, especially so in larger customers where many versions might be in play at any one time. Tracking of deployed versions would be important in the event of DR or cluster rebuilds where deploying an upgraded version might have detrimental impact on customers ability to operate their workloads.

But the versions (k8s/OS/CAPI) are defined in Cluster API CRs, so we will not lose them. Having anyway a `ClusterUpgrade` for each upgrade done can help in tracking changes sure (we can use `status` field) but we can find another approach and store upgrades in an independent DB in case we find using MC etcd is too much.